util/tls/tls_socket.cc

```diff
@@ -468,26 +562,161 @@ void TlsSocket::CancelOnErrorCb() {
+                 (red_res == Engine::NEED_WRITE)) {
+        HandleOpAsync(red_res);
+        // After handling, check if more decrypted data is available and continue loop.
+        continue;
+      } else {
+        // EOF_STREAM (connection closed probably), 0 (no data read) or unexpected error
+
+        break;
+      }
+    }
+  }
 }
 
 io::Result<size_t> TlsSocket::TrySend(io::Bytes buf) {
-  LOG(DFATAL) << "Not implemented";
-  return 0;
+  iovec v{const_cast<uint8_t*>(buf.data()), buf.size()};
+  return TrySend(&v, 1);
 }
 
 io::Result<size_t> TlsSocket::TrySend(const iovec* v, uint32_t len) {
-  LOG(DFATAL) << "Not implemented";
-  return 0;
+  size_t total_written = 0;
+  size_t iov_idx = 0;
+  size_t iov_offset = 0;
+  while (iov_idx < len) {
+    // Flush any pending output from previous iterations or previous calls.
+    while (engine_->OutputPending() > 0) {
+      auto ec = MaybeSendOutputAsync();
+      if (ec) {
+        if (ec == errc::resource_unavailable_try_again) {
+          // If we have already consumed some input in this call, return that amount.
+          // The caller will retry the rest.
+          if (total_written > 0) {
+            return total_written;
+          }
+          // If we haven't consumed anything, return EAGAIN so the caller waits.
+          return make_unexpected(ec);
+        }
+        VLOG(1) << "MaybeSendOutputAsync failed " << ec.message();
+        return make_unexpected(ec);
+      }
+      // If MaybeSendOutputAsync succeeded, it means we wrote something.
+      // We loop to check if there is more pending output.
+    }
+
+    // Create a temporary iovec (view) representing the unsent portion of the current buffer.
+    // This allows us to handle partial writes and retries without modifying the original
+    // iovec array.
+    iovec tmp;
+    tmp.iov_base = static_cast<uint8_t*>(v[iov_idx].iov_base) + iov_offset;
+    tmp.iov_len = v[iov_idx].iov_len - iov_offset;
+    PushResult push_res{PushToEngine(&tmp, 1)};
+    if (push_res.engine_opcode < 0) {
+      auto ec = HandleOpAsync(push_res.engine_opcode);
+      if (ec) {
+        if (ec == errc::resource_unavailable_try_again && total_written > 0) {
+          return total_written;
+        }
+        if (ec == errc::resource_unavailable_try_again) {
+          return make_unexpected(ec);
+        }
+        VLOG(1) << "HandleOpAsync failed " << ec.message();
+        return make_unexpected(ec);
+      }
+    }
+
+    if (push_res.written > 0) {
+      size_t written = push_res.written;
+      total_written += written;
+      if (written < tmp.iov_len) {
+        iov_offset += written;
+      } else {
+        ++iov_idx;
+        iov_offset = 0;
+      }
+      // We pushed data, so we likely have output.
+      // The next iteration of the loop will handle flushing it.
+      continue;
+    }
+  }
+
+  // This loop is necessary because the TLS engine may generate additional output after each flush,
```

Comment:
agree with the motivation but this can not be like this. it's not the  the async behavior.

Please add a test that has socket A writing into socket B, with B not reading. Continue with writes until A.upstream will start pushing back on writes to reproduce these cases.